container-search/src/main/java/com/yahoo/prelude/querytransform/QueryRewrite.java

```diff
@@ -156,53 +150,45 @@ private static Recall optimizeNotItemByRestrict(NotItem item, String restrictPar
         for (int i = item.getItemCount(); --i >= 1; ) {
             Item child = item.getItem(i);
             switch (optimizeByRestrict(child, restrictParam)) {
-            case RECALLS_EVERYTHING:
-                return Recall.RECALLS_NOTHING;
-            case RECALLS_NOTHING:
-                item.removeItem(i);
-                break;
+                case RECALLS_EVERYTHING:
+                    return Recall.RECALLS_NOTHING;
+                case RECALLS_NOTHING:
+                    item.removeItem(i);
+                    break;
             }
         }
         return Recall.UNKNOWN_RECALL;
     }
 
     private static Recall optimizeCompositeItemByRestrict(CompositeItem item, String restrictParam) {
+        Recall recall = Recall.UNKNOWN_RECALL;
         for (int i = item.getItemCount(); --i >= 0; ) {
             switch (optimizeByRestrict(item.getItem(i), restrictParam)) {
-            case RECALLS_EVERYTHING:
-                if ((item instanceof OrItem) || (item instanceof EquivItem)) {
-                    retainChild(item, i);
-                    return Recall.RECALLS_EVERYTHING;
-                } else if ((item instanceof AndItem) || (item instanceof NearItem)) {
-                    item.removeItem(i);
-                } else if (item instanceof RankItem) {
-                    // empty
-                } else {
-                    throw new UnsupportedOperationException(item.getClass().getName());
-                }
-                break;
-            case RECALLS_NOTHING:
-                if ((item instanceof OrItem) || (item instanceof EquivItem)) {
-                    item.removeItem(i);
-                } else if ((item instanceof AndItem) || (item instanceof NearItem)) {
-                    return Recall.RECALLS_NOTHING;
-                } else if (item instanceof RankItem) {
-                    item.removeItem(i);
-                } else {
-                    throw new UnsupportedOperationException(item.getClass().getName());
-                }
-                break;
+                case RECALLS_EVERYTHING:
+                    if ((item instanceof OrItem) || (item instanceof EquivItem)) {
+                        recall = Recall.RECALLS_EVERYTHING;
```

Comment:
Should you not check if term or field is of type filter ?

Reply:
Do you mean, keep removing them if isRanked is true?

Reply:
Correct.

Reply:
Ok, done.